### ISSUE TITLE
Transposed Convolution

### ISSUE THREAD

**Issue by ashimpd** [intel/mkl-dnn#193]
How do I achieve Transposed Convolution like in https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2DTranspose

**Comment by igorsafo** [comment]
Hi,

We are working on this functionality.
For now convolution primitive can be used to implement transposed convolution (swap forward and backward passes of a convolution).
The only problem is that backward pass of convolution doesn't support bias. if you want to reuse backward pass of convolution to compute forward pass of transposed convolution you will need to add bias by your own.

**Comment by ashimpd** [comment]
Thanks for the response. You mean I can use convolution_forward with prop_kind as "backward" during deployment.

**Comment by igorsafo** [comment]
Yes, regular convolution with prop kind "backward_data" can be used to calculate transposed convolution.

**Comment by ashimpd** [comment]
I tried this suggestion, however, I am not able to create the desc. Following is the code snippet

        memory::dims src_shape = { 1, 64, 31, 31 };
        memory::dims strides = {2, 2};
        memory::dims kernel = {3, 3};
        memory::dims dst_shape = {
            src_shape[0], // batch_size
            64,
            src_shape[2] * strides[0] + (kernel[0] - strides[0]),
            src_shape[3] * strides[1] + (kernel[1] - strides[1])
        };

        memory::dims weights_shape = {64, src_shape[1], kernel[0], kernel[1]};
        memory::dims padding_l = {0, 0};
        memory::dims padding_r = {0, 0};

        memory::desc src_md({src_shape}, memory::data_type::f32, memory::format::any);
        memory::desc dst_md({dst_shape}, memory::data_type::f32, memory::format::any);
        memory::desc weights_md({weights_shape}, memory::data_type::f32, memory::format::any);

        convolution_forward::desc conv_desc(prop_kind::backward_data,
                                            convolution_direct,
                                            src_md, weights_md,
                                            dst_md, strides, padding_l,
                                            padding_r, padding_kind::zero);


This gives following error
0x0000000000407e4f in mkldnn::error::wrap_c_api (
            status=mkldnn_invalid_arguments,
            message="could not create a convolution forward descriptor",
            error_primitive=0x0)

**Comment by vpirogov** [comment]
@ashimpd, what you are doing in the example is not supposed to work. For transposed convolution forward propagation is equivalent of backward propagation for regular convolution. So you can emulate transposed convolution using convolution_backward and prop_kind::backward_data.

**Comment by lzhengchun** [comment]
@vpirogov it would be nice if DNNL can provide an example to emulate transposed convolution.

**Comment by densamoilov** [comment]
@lzhengchun, we had a similar question before, please check if the way described in #143 works for you.

**Comment by vpirogov** [comment]
@lzhengchun, 

Deconvolution primitive in DNNL computes convolution gradient (or transposed convolution). The only difference between two is that deconvolution supports bias.

**Comment by lzhengchun** [comment]
Oh Thanks @vpirogov What I need is transposed convolution as **tf.nn.conv2d_transpose** implemented in TF. And TF said _it is actually the transpose (gradient) of conv2d rather than an actual deconvolution._ IF that is the case, DNNL's **deconv** should give the same results as TF's **tf.nn.conv2d_transpose** OK, problem goes back to issue #568

**Comment by vpirogov** [comment]
@lzhengchun TF actually uses older version of DNNL (v0.18 I assume) to implement this function, while your example uses v1.1. So I would expect that the results will be the same between your example and TF assuming all the parameters of computation are the same. Alok is trying to figure out where the difference comes from.

**Comment by lzhengchun** [comment]
Thanks @vpirogov meanwhile I can roll back to v0.18. Does V0.18 support Gen9?

**Comment by vpirogov** [comment]
You can try doing that as a part of debugging exercise. v0.18 is not API compatible with v1.1 and does not support GEN9. There's also no guarantee that the result will match TF computations as well.

**Comment by lzhengchun** [comment]
Thanks @vpirogov
